File: src/pml/sprites/validator.cpp
// ═══════════════════════════════════════════════════════════════════════════════
// PML Component Parameter Validator — Implementation
// ═══════════════════════════════════════════════════════════════════════════════

#include "validator.h"

#include <algorithm>
#include <cmath>
#include <iostream>

namespace pml {

// ═══════════════════════════════════════════════════════════════════════════════
// ParamSchema builder
// ═══════════════════════════════════════════════════════════════════════════════

ParamSchema& ParamSchema::enum_(const std::string& name,
                                const std::vector<std::string>& values,
                                const std::string& default_value) {
    FieldSpec spec;
    spec.type = FieldSpec::Enum;
    spec.enum_values = values;
    spec.default_value = default_value;
    m_fields[name] = std::move(spec);
    return *this;
}

ParamSchema& ParamSchema::number(const std::string& name,
                                 double default_value,
                                 double min_val,
                                 double max_val) {
    FieldSpec spec;
    spec.type = FieldSpec::Number;
    spec.default_value = default_value;
    spec.min_value = min_val;
    spec.max_value = max_val;
    m_fields[name] = std::move(spec);
    return *this;
}

ParamSchema& ParamSchema::boolean(const std::string& name, bool default_value) {
    FieldSpec spec;
    spec.type = FieldSpec::Boolean;
    spec.default_value = default_value;
    m_fields[name] = std::move(spec);
    return *this;
}

ParamSchema& ParamSchema::color(const std::string& name,
                                const std::string& default_value) {
    FieldSpec spec;
    spec.type = FieldSpec::Color;
    spec.default_value = default_value;
    m_fields[name] = std::move(spec);
    return *this;
}

ParamSchema& ParamSchema::any_type(const std::string& name,
                                   const Value& default_value) {
    FieldSpec spec;
    spec.type = FieldSpec::Any;
    spec.default_value = default_value;
    m_fields[name] = std::move(spec);
    return *this;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════════════

namespace {

[[nodiscard]] std::string value_to_str(const Value& v) {
    if (const auto* s = std::get_if<std::string>(&v)) return *s;
    if (const auto* sym = std::get_if<Symbol>(&v)) return sym->name;
    if (const auto* kw = std::get_if<Keyword>(&v)) return kw->name;
    return value_to_string(v);
}

[[nodiscard]] std::optional<double> value_to_number_opt(const Value& v) {
    if (const auto* i = std::get_if<int64_t>(&v)) {
        return static_cast<double>(*i);
    }
    if (const auto* d = std::get_if<double>(&v)) {
        return *d;
    }
    return std::nullopt;
}

}  // anonymous namespace

// ═══════════════════════════════════════════════════════════════════════════════
// validate_params
// ═══════════════════════════════════════════════════════════════════════════════

std::unordered_map<std::string, Value> validate_params(
    const ParamSchema& schema,
    const std::unordered_map<std::string, Value>& kwargs) {
    std::unordered_map<std::string, Value> result;

    for (const auto& [field_name, spec] : schema.fields()) {
        auto it = kwargs.find(field_name);
        if (it == kwargs.end() || std::holds_alternative<std::nullptr_t>(it->second)) {
            result[field_name] = spec.default_value;
            continue;
        }

        const Value& value = it->second;

        switch (spec.type) {
            case FieldSpec::Enum: {
                std::string s = value_to_str(value);
                const auto& allowed = spec.enum_values.value();
                bool found = std::find(allowed.begin(), allowed.end(), s) != allowed.end();
                if (found) {
                    result[field_name] = s;
                } else {
                    std::cerr << "Warning: invalid value '" << s << "' for '"
                              << field_name << "', falling back to '"
                              << value_to_str(spec.default_value) << "'\n";
                    result[field_name] = spec.default_value;
                }
                break;
            }
            case FieldSpec::Number: {
                if (auto n = value_to_number_opt(value)) {
                    double clamped = std::clamp(*n, spec.min_value, spec.max_value);
                    result[field_name] = clamped;
                } else {
                    result[field_name] = spec.default_value;
                }
                break;
            }
            case FieldSpec::Boolean: {
                if (const auto* b = std::get_if<bool>(&value)) {
                    result[field_name] = *b;
                } else {
                    result[field_name] = spec.default_value;
                }
                break;
            }
            case FieldSpec::Color: {
                if (std::holds_alternative<std::nullptr_t>(value)) {
                    result[field_name] = spec.default_value;
                } else {
                    result[field_name] = value_to_str(value);
                }
                break;
            }
            case FieldSpec::Any:
                result[field_name] = value;
                break;
        }
    }

    // Pass through unknown kwargs.
    for (const auto& [k, v] : kwargs) {
        if (result.find(k) == result.end()) {
            result[k] = v;
        }
    }

    return result;
}

}  // namespace pml

File: pml/sprites/validator.py
"""PML component parameter validator."""

from __future__ import annotations

from typing import Any

from pml.sprites.palette import resolve_color


class ParamSchema:
    """Schema for validating component parameters.

    Supports enum values (with fallback) and continuous values (with clamp).
    """

    def __init__(self) -> None:
        self._fields: dict[str, dict[str, Any]] = {}

    def enum(self, name: str, values: list[str], default: str) -> ParamSchema:
        """Register an enum parameter with a default fallback."""
        self._fields[name] = {
            "type": "enum",
            "values": values,
            "default": default,
        }
        return self

    def number(
        self, name: str, default: float, min_val: float = -1e9, max_val: float = 1e9
    ) -> ParamSchema:
        """Register a numeric parameter with clamp range."""
        self._fields[name] = {
            "type": "number",
            "default": default,
            "min": min_val,
            "max": max_val,
        }
        return self

    def boolean(self, name: str, default: bool = False) -> ParamSchema:
        self._fields[name] = {"type": "boolean", "default": default}
        return self

    def color(self, name: str, default: str = "#808080") -> ParamSchema:
        self._fields[name] = {"type": "color", "default": default}
        return self

    def any_type(self, name: str, default: Any = None) -> ParamSchema:
        self._fields[name] = {"type": "any", "default": default}
        return self

    def to_dict(self) -> dict[str, Any]:
        """Serialize schema to LLM-friendly dict format.

        Returns a dict with a "params" key containing a list of parameter specs:
        [{"name": ..., "type": ..., "default": ..., ...}, ...]
        """
        params: list[dict[str, Any]] = []
        for name, spec in self._fields.items():
            entry: dict[str, Any] = {
                "name": name,
                "type": spec["type"],
                "default": spec.get("default"),
            }
            if spec["type"] == "enum":
                entry["values"] = list(spec["values"])
            elif spec["type"] == "number":
                entry["min"] = spec["min"]
                entry["max"] = spec["max"]
            params.append(entry)
        return {"params": params}


def validate_params(schema: ParamSchema, kwargs: dict[str, Any]) -> dict[str, Any]:
    """Validate and clean keyword arguments against a schema.

    - Enum values: fallback to default if not in allowed set (with warning)
    - Numeric values: clamp to [min, max]
    - Missing values: use defaults
    - Unknown values: passed through (ignored by components that don't use them)
    """
    result: dict[str, Any] = {}

    for field_name, field_spec in schema._fields.items():
        value = kwargs.get(field_name)
        ftype = field_spec["type"]

        if value is None:
            result[field_name] = field_spec["default"]
            continue

        if ftype == "enum":
            if str(value) not in [str(v) for v in field_spec["values"]]:
                print(
                    f"Warning: invalid value '{value}' for '{field_name}', "
                    f"falling back to '{field_spec['default']}'"
                )
                result[field_name] = field_spec["default"]
            else:
                result[field_name] = str(value)

        elif ftype == "number":
            try:
                v = float(value)
                v = max(field_spec["min"], min(field_spec["max"], v))
                result[field_name] = v
            except (TypeError, ValueError):
                result[field_name] = field_spec["default"]

        elif ftype == "boolean":
            result[field_name] = bool(value)

        elif ftype == "color":
            raw = str(value) if value else field_spec["default"]
            result[field_name] = resolve_color(raw)

        else:
            result[field_name] = value

    # Pass through any extra kwargs not in schema
    for k, v in kwargs.items():
        if k not in result:
            result[k] = v

    return result
